primitives/runtime/src/offchain/pollable.rs

```diff
@@ -0,0 +1,439 @@
+//! on the runtime side (which is documented here).
+//!
+//! Most of the heavy lifting is done on the host side such as scheduling timers
+//! or performing network I/O, while the simple executor/reactor here is mostly
+//! to conveniently interface with this, enabling niceties like async/await
+//! syntax sugar or future combinators.
+//!
+//! Because of this approach, the following `Future`s are supported here:
+//! 1. immediately ready ones (i.e. which return `Poll::Ready`),
+//! 2. internally host-driven ones via `PollableId`,
+//! 3. combinators for the supported futures.
+
+// Based on https://github.com/tomaka/redshirt/blob/802e4ab44078e15b47dc72afc6b7e45bab1b72df/interfaces/syscalls/src/block_on.rs#L88.
+
+#![allow(dead_code, unused_variables)]
+
+use sp_core::offchain::{PollableId, PollableKind};
+use sp_core::offchain::{HttpError, HttpRequestId, HttpRequestStatus};
+use sp_std::{rc::Rc, collections::btree_set::BTreeSet};
+use sp_std::prelude::*;
+
+use core::cell::RefCell;
+use core::convert::{TryFrom, TryInto};
+use core::future::Future;
+use core::mem::{self, ManuallyDrop};
+use core::pin::Pin;
+use core::task::{self, Context, Poll, Waker, RawWaker, RawWakerVTable};
+
+use spin::Mutex;
+use pin_project::pin_project;
+
+pub(crate) type MessageId = PollableId;
+
+/// Pins a value on the stack.
+// A copy of `futures::pin_mut!` without having to pull `futures` in the no_std
+// context.
+macro_rules! pin_mut {
+	($($x:ident),* $(,)?) => { $(
+		// Move the value to ensure that it is owned
+		let mut $x = $x;
+		// Shadow the original binding so that it can't be directly accessed
+		// ever again.
+		#[allow(unused_mut)]
+		let mut $x = unsafe {
+			core::pin::Pin::new_unchecked(&mut $x)
+		};
+	)* }
+}
+
+fn epoll_peek(interest_list: &[MessageId]) -> Option<MessageId> {
+	let deadline = sp_io::offchain::timestamp();
+	sp_io::offchain::pollable_wait(interest_list, Some(deadline))
+}
+
+fn epoll_wait(interest_list: &[MessageId]) -> MessageId {
+	sp_io::offchain::pollable_wait(interest_list, None)
+		.expect(
+			"pollable_wait with None deadline should block indefinitely \
+			until we get a response; qed"
+		)
+}
+
+pub(crate) fn next_notification(interest_list: &[MessageId], block: bool) -> Option<(MessageId, usize)> {
+	let id = if !block {
+		epoll_peek(interest_list)
+	} else {
+		Some(epoll_wait(interest_list))
+	}?;
+
+	let pos = interest_list.iter().position(|&x| x == id)
+		.expect("pollable API should only return an ID from the interest list; qed");
+
+	Some((id, pos))
+}
+
+/// Registers a message ID and a waker. The `block_on` function will
+/// then ask the kernel for a message corresponding to this ID. If one is received, the `Waker`
+/// is called.
+///
+/// For non-interface messages, there can only ever be one registered `Waker`. Registering a
+/// `Waker` a second time overrides the one previously registered.
+pub(crate) fn register_message_waker(message_id: MessageId, waker: Waker) {
```

Comment:
Can we move this closer to `HostFuture`? This seems like an implementation detail of that future imho. Also does it have to be `pub`?